jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -1566,18 +1566,24 @@ private void completeChannelCallback(Throwable failure)
                 stream = httpChannelState._stream;
                 assert httpChannelState._callbackFailure == null;
 
-                // Turn pending demand or unconsumed input on persistent connections into failure
+                // Turn pending demand into failure.
                 if (httpChannelState._onContentAvailable != null)
+                {
                     failure = ExceptionUtil.combine(failure, new IllegalStateException("demand pending"));
-                else if (httpChannelState.getConnectionMetaData().isPersistent())
+                }
+                // Else failure for a persistent non expecting request to not consume the request
+                else if (httpChannelState.getConnectionMetaData().isPersistent() && !httpChannelState._expects100Continue)
                     failure = ExceptionUtil.combine(failure, stream.consumeAvailable());
```

Comment:
Add braces.

But then again, this was the fix @lorban made in #13934, as `stream.consumeAvailable()` changes the persistent state.

And `consumeAvailable()` is called even when the connection is not persistent, contrary to the comment you made in #13934.

I think #13934 was more readable and had less special cases, with the addition to skip `consumeAvailable()` if not persistent.